File: app/models/listing.rb
class Listing < ActiveRecord::Base
  belongs_to :course
  belongs_to :term
  has_many   :sections, dependent: :destroy
  validates  :longname, presence: true

  def credits
    min_credits == max_credits ? "#{min_credits}" : "#{min_credits}-#{max_credits}"
  end
end

File: packages/yacs-api-client/src/lib/listing.ts
import {
  Model,
  SpraypaintBase,
  Attr,
  BelongsTo,
  HasMany
} from "spraypaint";

import { ApplicationRecord } from "./application-record";
import { Course } from "./course";
import { Term } from "./term";
import { Section } from "./section";

@Model()
export class Listing extends ApplicationRecord {
	static jsonapiType = "listings";

	@Attr() tags: string[];
	@Attr() active: boolean;
	@Attr() maxCredits: number;
	@Attr() minCredits: number;
	@Attr() description: string;
	@Attr() longname: string;
	@Attr() courseShortname: string;
	@Attr() subjectShortname: string;
	@Attr() uuid: string;

	@BelongsTo() course: Course;
	@BelongsTo() term: Term;
	@HasMany() sections: Section[];
}
